Fixes #23765 - bcrypt hash password instead sha1

lzap opened the pull request:
This is transparent implementation of bcrypt hashing which replaces
sha1. It allows admin to set cost value as a number. The implementation
is transparent - it just works for all users who change their password
after the upgrade. Current users can keep using sha1 mechanism until
they change passwords.

Here is how it works:

```
[5] pry(main)> u = User.second
=> #<User:0x0000555560a1f468
 id: 4,
 login: "admin",
 firstname: "Admin",
 lastname: "User",
 mail: "root@brq.redhat.com",
 admin: true,
 last_login_on: Thu, 31 May 2018 10:45:20 UTC +00:00,
 auth_source_id: 1,
 created_at: Thu, 01 Jun 2017 10:13:01 UTC +00:00,
 updated_at: Thu, 01 Jun 2017 10:14:14 UTC +00:00,
 password_hash: "38a7c6088e13799b4dd70d76e2bfeb5720ea360e",
 password_salt: "4852ef8b3e0388852f19e283af278b8a3c6327e5",
 locale: nil,
 avatar_hash: nil,
 default_organization_id: nil,
 default_location_id: nil,
 lower_login: "admin",
 mail_enabled: true,
 timezone: nil,
 description: nil>
[6] pry(main)> u.password = "changeme"
=> "changeme"
[7] pry(main)> u.save!
aud|I update event for User with id 4
per|D verifying the transaction by permission edit_users for class User
=> true
[8] pry(main)> User.second
=> #<User:0x0000555559e48aa0
 id: 4,
 login: "admin",
 firstname: "Admin",
 lastname: "User",
 mail: "root@brq.redhat.com",
 admin: true,
 last_login_on: Thu, 31 May 2018 10:45:20 UTC +00:00,
 auth_source_id: 1,
 created_at: Thu, 01 Jun 2017 10:13:01 UTC +00:00,
 updated_at: Fri, 01 Jun 2018 07:15:16 UTC +00:00,
 password_hash: "$2a$10$7qC1rkwPYt6xyXV5o/sj9.UNIq/Rh/XXRFsA99fmweko2aKggnGl.",
 password_salt: "$2a$10$7qC1rkwPYt6xyXV5o/sj9.",
 locale: nil,
 avatar_hash: nil,
 default_organization_id: nil,
 default_location_id: nil,
 lower_login: "admin",
 mail_enabled: true,
 timezone: nil,
 description: nil>
[9] pry(main)> u.matching_password?("changeme")
=> true
[10] pry(main)> u.matching_password?("blahblah")
=> false
```

It introduces new dependency, no other 2nd level deps tho.

lzap committed:
Fixes #23765 - bcrypt hash password instead sha1

lzap commented:
Fixed tests, bcrypt salt must be in expected format starting with `$2`. The current implementation actually copies salt into password field, but that's fine - we designed our user table in a way that salt and password hash are separate columns, this allows using hash functions which keep them separate (like SHA).

lzap commented:
And it broke lots of tests :-) Will fix on Monday. Everytime a `Setting` is mocked, it breaks. This needs a helper which will stub settings in a nicer way.

ekohl commented:
Many systems allow transparant upgrading of hashing on login. Might be something to consider implementing.

ik5 commented on test/fixtures/settings.yml:
```diff
@@ -384,3 +384,8 @@ attribute80:
+attribute80:
+  name: bcrypt_cost
+  category: Setting::Auth
+  default: 10
```
why 10, isn't it a bit on the lower side (if it's between 5-30)?

lzap replied:
If you set this to 30, test suite will not finish in an hour on desktop CPU 2017 (I cancelled it), I think we are on safe side :-)

lzap replied:
After reading some material it looks like 10 should be safe but setting it to 11-12 should not be a problem for todays computers (hashing up to 300ms in the worst case). I think I am gonna increase this to 11 or 12 and I will add telemetry histogram which will report hashing duration so admin can decide herself.

https://security.stackexchange.com/questions/17207/recommended-of-rounds-for-bcrypt

ik5 replied:
> If you set this to 30, test suite will not finish in an hour on desktop CPU 2017 (I cancelled it), I think we are on safe side :-)

I know that 30 is expensive, thanks for explaining the "10" part.

ik5 reviewed (commented):
While I think that salting is better then just a hash (and sha2 is better then sha1), I have mix feeling about bcrypt.

It solves the 90's and 2000's issues, but today it is not a match for a strong ARM64 bit machine.

I prefer using `bcrypt` that solves the issues that `bcrypt` still have:
https://medium.com/@mpreziuso/password-hashing-pbkdf2-scrypt-bcrypt-1ef4bb9c19b3

lzap commented:
> Many systems allow transparant upgrading of hashing on login. Might be something to consider implementing.

That make sense for services with thousands of users, but typical foreman instance has up to 100 if I am not mistaken. While I agree and it's good idea, I'd rather keep this patch small.

lzap commented:
> It solves the 90's and 2000's issues, but today it is not a match for a strong ARM64 bit machine.
> I prefer using bcrypt that solves the issues that bcrypt still have:

These to sentences do not make sense to me, can you elaborate please?

lzap commented:
Before we decide, let's compare:

```ruby
require 'benchmark'
require 'bcrypt'
password = 'goats are great!'
Benchmark.bm do |bench|
  for cost in 6..20
    bench.report("Cost factor: #{cost}\t") do
      BCrypt::Password.create(password, cost: cost)
    end
  end
end
```

Intel i7–4770 from 2013:

```
                      user     system      total        real
Cost factor: 6    0.010000   0.000000   0.010000 (  0.003622)
Cost factor: 7    0.010000   0.000000   0.010000 (  0.007590)
Cost factor: 8    0.010000   0.000000   0.010000 (  0.013918)
Cost factor: 9    0.030000   0.000000   0.030000 (  0.028723)
Cost factor: 10   0.060000   0.000000   0.060000 (  0.055023)
Cost factor: 11   0.100000   0.000000   0.100000 (  0.110837)
Cost factor: 12   0.230000   0.000000   0.230000 (  0.226982)
Cost factor: 13   0.450000   0.000000   0.450000 (  0.446507)
Cost factor: 14   0.890000   0.000000   0.890000 (  0.892617)
Cost factor: 15   1.770000   0.000000   1.770000 (  1.778647)
Cost factor: 16   3.580000   0.000000   3.580000 (  3.588397)
Cost factor: 17   7.130000   0.000000   7.130000 (  7.155973)
Cost factor: 18  14.510000   0.010000  14.520000 ( 14.572114)
Cost factor: 19  28.720000   0.020000  28.740000 ( 28.860461)
Cost factor: 20  57.440000   0.000000  57.440000 ( 57.623139)
```

AMD Ryzen 1700 from 2017:
```
Cost factor: 6    0.006218   0.000000   0.006218 (  0.006249)
Cost factor: 7    0.007863   0.000000   0.007863 (  0.007878)
Cost factor: 8    0.015588   0.000000   0.015588 (  0.015616)
Cost factor: 9    0.031011   0.000000   0.031011 (  0.031071)
Cost factor: 10   0.061754   0.000000   0.061754 (  0.061865)
Cost factor: 11   0.123333   0.000000   0.123333 (  0.123545)
Cost factor: 12   0.247307   0.000000   0.247307 (  0.247857)
Cost factor: 13   0.493634   0.000000   0.493634 (  0.494498)
Cost factor: 14   0.987151   0.000000   0.987151 (  0.988881)
Cost factor: 15   1.975219   0.000000   1.975219 (  1.978611)
Cost factor: 16   3.954145   0.000000   3.954145 (  3.960999)
Cost factor: 17   7.922481   0.000000   7.922481 (  7.937260)
Cost factor: 18  15.824174   0.000000  15.824174 ( 15.851681)
Cost factor: 19  31.633856   0.000000  31.633856 ( 31.695432)
Cost factor: 20  63.218303   0.000542  63.218845 ( 63.333387)
```

ik5 commented:
> >It solves the 90's and 2000's issues, but today it is not a match for a strong ARM64 bit machine.
I prefer using bcrypt that solves the issues that bcrypt still have:

> These to sentences do not make sense to me, can you elaborate please?

bcrypt uses blowfish algorithm and it has known [weakness](https://en.wikipedia.org/wiki/Blowfish_(cipher)#Weakness_and_successors) that can help me "open" the key that is generated faster.

So due to that issue, bcrypt usualy truncate some of the text length, usually between 50-70 bytes long, and that another way to help "guess" the password, because you know the maximum length that it used, so a password longer then 71 will be "only"70 chars.

The CPU cost for it can handle by a single CPU this days, something that when it was designed, it costs a lot more machines then you need at the current time. So a single CPU today holds the power to calculate the weakness in blowfish that even 5 years ago needed more then a single CPU.

At least to my knowledge in the matter, and I might be wrong in my understanding here :)

ekohl commented:
> That make sense for services with thousands of users, but typical foreman instance has up to 100 if I am not mistaken. While I agree and it's good idea, I'd rather keep this patch small.

Oh I agree it's out of scope for this PR.

lzap commented:
I don't want to start discussion about if bcrypt is weak or not here, in general it's one of the recommended password hashing methods and that should be enough for us. Let's take this to some crypto mailing lists instead :-)

But thanks for the default cost point, it was really useful. After reading https://labs.clio.com/bcrypt-cost-factor-4ca0a9b03966 I implemented *calibration* via seed script: default cost is set to be at least 11 and as high as CPU can calculate it in 250ms. On my CPU this is currently cost 12, but on every seed (e.g. Foreman upgrade) this will be calibrated to be at least 250ms.

ik5 commented:
okay, fair enough :)

tbrisker commented:
2 questions come to mind here, let me know if i'm completely off:
1. does this mean every api call that needs to verify password will take 250ms longer? this might not be too bad for a single request, but for certain cases it would potentially have a very significant affect (think for example of the ansible inventory script that does a separate call for every host - with thousands of hosts this means ~4min/1k hosts difference). this might be possible to mitigate with tokens or something, but should be considered
2. Should we be in the business of worrying about authentication for higher-security environment at all? or should the best practice in these cases be "use external authentication/ldap" which allows better control of various configurations that can enhance security?

ik5 commented:
@tbrisker my opinion on your questions:
1. You can send the API with session to my understanding, so it will not check authentication every time.
2. It's always good practice to take care of it, in my opinion.

lzap commented:
1. Yes, you are right. Most of our users don't use API sessions, at least our installer configures hammer with stateless username/password and this is right concern. Perhaps 250ms is too much for a default. Looks like SHA1 calculates on my system under 1ms (not measurable), looks like cost factor of 6 is 4ms on my system.

Therefore I am changing the calibration threshold to 5ms which is very performance-safe default, it will be cost around 5-8 depending on the CPU power in 2018. Users who need security can increase this to any cost they want.

2. Yes, we should care. We have many users deploy Foreman in public clouds without LDAP. SHA-1 is already today quite fast and cracking password is easy. We should be able to let users to respond to this and bcrypt is exactly the answer. It allows us to recalibrate for the future while with SHA1 or similar algorithms we'd need to reevaluate and refactor every few years.

```
       user     system      total        real
SHA1              0.000228   0.000000   0.000228 (  0.000231)
Cost factor: 1    0.001192   0.000000   0.001192 (  0.001199)
Cost factor: 2    0.001127   0.000000   0.001127 (  0.001132)
Cost factor: 3    0.001126   0.000000   0.001126 (  0.001131)
Cost factor: 4    0.001131   0.000008   0.001139 (  0.001145)
Cost factor: 5    0.002079   0.000020   0.002099 (  0.002106)
Cost factor: 6    0.004017   0.000000   0.004017 (  0.004031)
Cost factor: 7    0.007853   0.000018   0.007871 (  0.007898)
Cost factor: 8    0.015571   0.000000   0.015571 (  0.015615)
Cost factor: 9    0.030963   0.000026   0.030989 (  0.031087)
Cost factor: 10   0.061681   0.000031   0.061712 (  0.061859)
Cost factor: 11   0.123340   0.000026   0.123366 (  0.123637)
Cost factor: 12   0.246535   0.000047   0.246582 (  0.247097)
```

ik5 commented:
@lzap, have you seen the following new PR?
https://github.com/theforeman/foreman/pull/5758
https://github.com/theforeman/foreman/pull/5759

lzap commented:
Rebased conflict, fixed tests.

lzap commented:
@ik5 that's something not relevant, the other PRs are about hashing algo for OS installer (like root passwords etc but thanks for the poke :-) Rebased.

ik5 commented:
oh, then n/m :)

ik5 commented:
went over the code again after the rebase, nothing looks different in my opinion, so lgtm

tbrisker commented:
I'm not an expert, but isn't cost<10 considered not secure? how do we know that the cost selected is actually more secure then sha and we aren't actually reducing security here?

ik5 commented:
@tbrisker sha without salt is insecure, the salt makes it harder to handle.
bcrypt has it's issues as well, not only the complexity but also the password length can be a factor.
A password longer then 70 chars (or something like that) is not suited for it and other mechanisms such as scrypt for example.

having said that lower cpu cycle complexity then 10 is not a good advice imo.

lzap commented:
From the table above, I can tell that SHA takes `0.000228` sec and bcrypt with cost 6 is `0.004017`. Therefore I can say that bcrypt is slower (read: harder to crack) by order of one magnitude on my system. This matches what can be googled on this topic - bcrypt is considered more safe than SHA/MD5 "naive" hashing.

Yes, cost lower than 10 is not recommended but it is still an improvement for us. Users who want to secure their system can increase this value to reasonable level, we warn them that this will affect performance of session-less API. Also the calibration will make sure that Foreman is hashing on reasonable level of security.

This is a tradeoff between security and performance of login and there is no best cost value. I think than we can afford extra 4 miliseconds by default giving users ability to tighten it more if they want. There is no good value I can offer you guys. Let me start a thread about it: https://community.theforeman.org/t/foreman-bcrypt-password-default-cost/10343

ekohl commented on db/seeds.d/010-calibrate_bcrypt_cost.rb:
```diff
@@ -0,0 +1,11 @@
+# calibrate default cost to be at least 5ms (performance-wise safe)
```
This whole block should be behind an `if SETTINGS[...]` so admins can always override this. This also makes it easy to use a cost of 1 in development mode for example.

lzap replied:
In test environment, cost is set to 1 automatically. I kinda like to have development environment as close as production if it makes sense and here I think that it does. If there was regression in calibration code, we would immediately see this if that code is same. Not sure if 1 or 5 ms makes things any better during development.

ekohl commented on db/seeds.d/010-calibrate_bcrypt_cost.rb:
```diff
@@ -0,0 +1,11 @@
+# calibrate default cost to be at least 5ms (performance-wise safe)
+new_cost = BCrypt::Engine.calibrate(5)
```
Perhaps the user should be able to override the 5 here? I've always been taught to at least use constants for magic values or make them configurable.

lzap replied:
I lean towards less complex interface for users. Now there is just "cost" value exposed, you want to expose "cost" and also "calibration time" which actually sets "cost" and it's hard to explain. I'd rather keep this as constant, user can always increase cost and the seed will not touch it anymore.

ekohl commented on db/seeds.d/010-calibrate_bcrypt_cost.rb:
```diff
@@ -0,0 +1,11 @@
+new_cost = BCrypt::Engine.calibrate(5)
+
+# keep minimum reasonable cost in case CPU was busy
+new_cost = 4 if new_cost < 4
```
Same here with 4.

lzap replied:
Same concern - you want me to expose now three numbers, that's a challenge to explain for advanced users and impossible for noobs - they will have no idea how to set this and they will set it wrong.

Constants are for good, please don't make it dogma to expose them all that's not possible. This one is a security check to avoid setting cost to 1 in case CPU was busy during the short test. Cost values 1-4 can be calculated under 1 ms. I can however remove this check, having CPU so slow that it would end up being cost 1 is highly unlikely.

lzap commented:
Rebased so the calibration sets default cost to burn at least 50ms which is cost 10 on my modern CPU from 2017. Let's just merge this and if we will have anybody complaining about slow session-less API, workarounds are trivial (use a session or change password via Rake Console with lower cost).

lzap commented:
Jenkins error, any other comments before I rebase once again?

timogoebel commented:
@lzap: Can we also have this for [Personal Access Tokens](https://github.com/theforeman/foreman/blob/develop/app/models/personal_access_token.rb)?

lzap commented:
Sure, done. Rebased.

lzap commented:
Guys, I've spent good amount of time on this, I really did my homework testing all of this. I'd appreciate re-review, I don't see any reason why not to go with this.

Rebased to re-trigger tests. I saw some failures, need to investigate but take a look on the current code and let me know if I should proceed with fixing tests and final rebase. Thanks.

timogoebel commented on test/factories/personal_access_tokens.rb:
```diff
@@ -1,8 +1,8 @@
     sequence(:name) {|n| "Personal Access Token #{n}" }
-    token { Digest::SHA1.hexdigest(SecureRandom.urlsafe_base64(nil, false)) }
     revoked { false }
+    token { BCrypt::Engine.hash_secret('OSKoHLYqOj0AQd5dnyb8sw', PersonalAccessToken.token_salt_bcrypt(user.id)) }
```
Where does `user` come from?

lzap replied:
Whoops that was the last rebase, let me take a look. But please keep going!

lzap replied:
Fixed, let's use a constant token for mock.

timogoebel commented on app/models/personal_access_token.rb:
```diff
@@ -19,23 +20,48 @@ class PersonalAccessToken < ApplicationRecord
-    Digest::SHA1.hexdigest([token, token_salt(user)].join)
+  def self.hash_token(user, token, method = :bcrypt)
+    if method == :bcrypt
+      telemetry_duration_histogram(:login_pwhash_duration, :ms, algorithm: method) do
```
Can we move this one level up?

lzap replied:
Right, it's now much more simple.

timogoebel commented on app/models/personal_access_token.rb:
```diff
@@ -19,23 +20,48 @@ class PersonalAccessToken < ApplicationRecord
-  def self.hash_token(user, token)
-    Digest::SHA1.hexdigest([token, token_salt(user)].join)
+  def self.hash_token(user, token, method = :bcrypt)
+    if method == :bcrypt
```
I believe this would look a lot cleaner of we'd extract this to two classes.

Just to give you an idea (not perfect, yet)

```ruby
def self.hash_token(user, token, method = :bcrypt)
  hash_implementation(method).hash_token(user, token)
end

def self.hash_implementation(method = :bcrypt)
   method == :sha ? ShaHash : BcryptHash
end

class ShaHash
   def hash_token(payload, token)
      ...
   end

   def token_salt(payload)
     ...
   def
end
```

We could then reuse this for the `User` model and test it properly.

lzap replied:
Good idea, implemented whole new class.

timogoebel commented on test/models/user_test.rb:
```diff
@@ -930,6 +930,14 @@ def editing_self_helper
 
+  test "#matching_password? succeeds if password matches with higher cost" do
+    Setting[:bcrypt_cost] = 15
+    u = FactoryBot.build_stubbed(:user)
```
nit: `user`

lzap replied:
Sure.

timogoebel reviewed (commented):
I think this increases security and that's super important for such a critical tool like Foreman. So, 👍 in general.

lzap commented:
Rebased, it's not much more clean code.

lzap commented:
And packaging PR is: https://github.com/theforeman/foreman-packaging/pull/3487

timogoebel reviewed (dismissed):
Code looks good to me. This still needs manual testing.

ezr-ondrej commented on Gemfile:
```diff
@@ -50,6 +50,7 @@ gem 'mail', '~> 2.7'
 gem 'sshkey', '~> 1.9'
 gem 'dynflow', '>= 1.0.0', '< 2.0.0'
 gem 'daemons'
+gem 'bcrypt'
```
I would prefer to close the dependency `'~> 3.1'`, if there is no reason not to.

lzap replied:
My thinking was the API will hardly change but sure not a big deal. Lemme do it.

ezr-ondrej commented on lib/foreman/password_hash.rb:
```diff
@@ -0,0 +1,63 @@
+      end
+
+      def calculate_salt(object, cost)
+        "$2a$#{cost.to_s.rjust(2, '0')}$#{Digest::SHA1.hexdigest(object.to_s)}"
```
Would be possible to throw the calculation? Meaning call `generate_salt` in here, to get less dependent on our own implementation for the future?

lzap replied:
Sorry what you mean? Throw away? For some reason @timogoebel designed the Personal Tokens in a way that salt is a digest of user ID and I wanted to keep this behavior.

timogoebel replied:
I believe I just copied the behavior from the user model. But I don't remember.

tbrisker replied:
wait, isn't the whole idea of using a salt based on it being random? if it is derived from the object, it's not random anymore...

timogoebel replied:
I believe it does not have to be random. It just has to be unique per user.

ezr-ondrej replied:
I believe it is not about being random, but mainly being different, so the password hashes are not similar to hashes of the same password taken elsewhere (or in the same database). Thus I believe the idea is it should be unique per password, not user. But I think being unique per user is safe enough, because we got just one real password (otherwise i think it would be a problem - consider the user is setting same password twice, on both places - the hashes are being same and it gives an attacker first clue about the user's password).

ezr-ondrej commented on lib/foreman/password_hash.rb:
```diff
@@ -0,0 +1,63 @@
+    end
+
+    delegate :generate_salt, :calculate_salt, :hash_secret, to: :implementation
+
```
I quite like the syntax of https://github.com/codahale/bcrypt-ruby/blob/v3.1.12/lib/bcrypt/password.rb (meaning the #==, .create and .new), you think would be worth a while to implement it as well?

lzap replied:
I totally hate this ;-) The thing is it makes the code unreadable, you expect the `==` operator to do something while it secretly does something else. I like Ruby's flexibility to build DSL where it's obvious it's not regular Ruby but this is a sneaked-in operator and when it comes to security this can be pretty darn dangerous.

I am strongly for being explicit in this case.

ezr-ondrej replied:
Yeah, reading this, makes sence :+1:

ezr-ondrej commented on app/models/user.rb:
```diff
@@ -554,7 +557,7 @@ def authenticate_by_personal_access_token(token)
   def prepare_password
     if password.present?
-      self.password_salt = Digest::SHA1.hexdigest([Time.now.utc, rand].join)
+      self.password_salt = salt_password
```
We are storing the salt twice for the new implementation, right? Would it be safe to assume we are using new (bcrypt) implementation for all new passwords and thus leave the salt column blank?
It would help with the `detect_implementation` as well :+1:

lzap replied:
I like it, let me think about it.

lzap replied:
Hmm you know what? As much as I would like to have the `password_salt` empty or nil, it makes the code in User model class unnecessary complex. I need to test for `password_salt` existence on multiple places as SHA1 is built around salt separated from the password. I cannot take the nice approach of defining some kind of Password model because this will not work for SHA1. I dunno, will take a look in the morning if I am able to come up with something that looks nice. I might stick with simple copy.

ezr-ondrej replied:
if it would make the code less readable, I believe it is not necessary, we are still being able to guess the algorithm used safely.

lzap replied:
It's probably not worth the effort.

timogoebel commented:
This is still waiting on packaging ACK.

ezr-ondrej reviewed (commented):
Code looks good, just few thoughts of mine, but feel free to toss them, no merge blockers. :man_shrugging: :wink: 
I have tested this:
- [passed] logged with user created before change
- [passed] changed password for existing user and logged by it
- [passed] created user with new algorithm and logged to the account.

lzap commented:
@timogoebel packaging PR merged. I will do one another rebase tho to addres @ezr-ondrej notes.

tbrisker commented on app/models/setting/auth.rb:
```diff
@@ -24,6 +24,7 @@ def self.default_settings
       self.set('authorize_login_delegation', N_("Authorize login delegation with REMOTE_USER environment variable"), false, N_('Authorize login delegation')),
       self.set('authorize_login_delegation_api', N_("Authorize login delegation with REMOTE_USER environment variable for API calls too"), false, N_('Authorize login delegation API')),
       self.set('idle_timeout', N_("Log out idle users after a certain number of minutes"), 60, N_('Idle timeout')),
+      self.set('bcrypt_cost', N_("Cost value of bcrypt password hash function for internal auth-sources (4-30). Higher value is safer but verification is slower particularly for stateless API calls and UI logins. Password change needed to take effect."), 4, N_('BCrypt password cost')),
```
I wish we can handle this without adding yet another setting by choosing a reasonable default....

timogoebel replied:
Is this a blocker for you, @tbrisker?

lzap replied:
This should be configurable, I don't think we should be opinionated on security-related things. We actually set a reasonable value during seed, but I'd probably suggest to bump the cost to more safe number like 11 or 12.

I could not reply in the previous thread, so let me reply here:

> wait, isn't the whole idea of using a salt based on it being random? if it is derived from the object, it's not random anymore...

Yes, but for Personal Token the token is actually a random stuff so it does not matter. Honestly I did not investigate why @timogoebel designed it this way, but I can confirm this is safe and I did not want to change behavior in this patch.

lzap replied:
@tbrisker as much as I would love to provide sane default, re-read our dicussion on discourse about this. Every single person a different view on how the cost should be set. So we default to 50ms processing time during calibration as a reasonably safe but also fast time and we let users to decide.

ekohl replied:
I also disagree that every setting is bad. IMHO settings should have sane defaults, but being able to tweak the app can be a very good thing.

timogoebel commented:
Another thought came to my mind: When a user logs in, we see the password in clear text. Do we want to use that to update the password stored in the database from sha to bcrypt? We don't have to do it in this PR, I believe we're close to merging. Thoughts?

ezr-ondrej commented:
@timogoebel I like the idea! Anyway I would not push it to this PR. But sure sounds nice, it would make it more likely to migrate fully in some distant future.

lzap commented:
Since there is no other way of upgrading people to bcrypt other than in the moment of login, I agree. Let me implement that.

lzap commented:
Amended two changes. First, added version number in the `Gemfile`. Second, added upgrade mechanism for existing users.

```diff
diff --git a/app/models/user.rb b/app/models/user.rb
index 5b2020d17..3562df5f1 100644
--- a/app/models/user.rb
+++ b/app/models/user.rb
@@ -325,10 +325,19 @@ class User < ApplicationRecord
     User.where(:lower_login => userlist.map(&:downcase)).pluck(:id)
   end
 
+  def upgrade_password(pass)
+    logger.info "Upgrading password for user #{self.login} to bcrypt"
+    self.password_salt = salt_password
+    self.password_hash = hash_password(pass)
+    update_columns(password_salt: self.password_salt, password_hash: self.password_hash)
+  end
+
   def matching_password?(pass)
     return false if self.password_hash.nil?
     type = Foreman::PasswordHash.detect_implementation(self.password_salt)
-    self.password_hash == hash_password(pass, type)
+    matches = self.password_hash == hash_password(pass, type)
+    upgrade_password(pass) if type != :bcrypt && matches
+    matches
   end
 
   def my_usergroups
diff --git a/test/models/user_test.rb b/test/models/user_test.rb
index 7abe1d3d0..e84fac354 100644
--- a/test/models/user_test.rb
+++ b/test/models/user_test.rb
@@ -944,6 +944,24 @@ class UserTest < ActiveSupport::TestCase
     refute u.matching_password?('wrong password')
   end
 
+  test "#matching_password? upgrades from SHA1 to BCrypt" do
+    hasher = Foreman::PasswordHash.new(:sha1)
+    u = FactoryBot.build_stubbed(:user)
+    u.password_salt = hasher.generate_salt(0)
+    u.password_hash = hasher.hash_secret('password', u.password_salt)
+    u.expects(:upgrade_password).with('password')
+    assert u.matching_password?('password')
+  end
+
+  test "#matching_password? does not upgrade from BCrypt to BCrypt for no reason" do
+    hasher = Foreman::PasswordHash.new(:bcrypt)
+    u = FactoryBot.build_stubbed(:user)
+    u.password_salt = hasher.generate_salt(5)
+    u.password_hash = hasher.hash_secret('password', u.password_salt)
+    u.expects(:upgrade_password).never
+    assert u.matching_password?('password')
+  end
+
   test ".except_hidden doesn't return any hidden users" do
     assert User.unscoped.where(:auth_source_id => AuthSourceHidden.first).any?
     User.unscoped.except_hidden.each do |user|
```

lzap commented:
Test failure unrelated I think, but I am rebasing. This is ready for merge.

timogoebel commented on app/models/user.rb:
```diff
@@ -324,8 +325,19 @@ def self.fetch_ids_by_list(userlist)
-    self.password_hash == hash_password(pass)
+    return false if self.password_hash.nil?
+    type = Foreman::PasswordHash.detect_implementation(self.password_salt)
+    matches = self.password_hash == hash_password(pass, type)
```
Can we return early here just to be sure we don't overwrite the password hash?

```ruby
return false if self.password_hash != hash_password(pass, type)
upgrade_password(pass) if type != :bcrypt
true
```

timogoebel commented:
@lzap: I just have one small comment regarding the code. Otherwise: This is ready to merge. I tested this a bit and it works very well.

lzap commented:
Sure, rebased.

timogoebel commented:
@lzap: Mind taking a look at the test failures?

lzap commented:
Sure, done.

timogoebel commented:
Katello tests failures are related, but imho this is good to merge.

timogoebel commented:
Katello PR: https://github.com/Katello/katello/pull/8091
Packaging PR: https://github.com/theforeman/foreman-packaging/pull/3487

@ekohl, @evgeni: Can you ack and merge, please?

lzap commented:
https://community.theforeman.org/t/tutorial-upgrading-passwords-to-bcrypt/13745

mmoll commented:
merged, díky @lzap!
